src/java.base/share/classes/java/security/Security.java

```diff
@@ -74,11 +90,215 @@ private static class ProviderProperty {
+            } catch (InvalidPathException e) {
+                return e;
+            }
+            loadFromPath(path, mode);
+            return null;
+        }
+
+
+        private static Exception loadExtraFromFileUrl(URI uri, LoadingMode mode)
+                throws Exception {
+            Path path;
+            try {
+                path = Path.of(uri);
+            } catch (Exception e) {
+                return e;
+            }
+            loadFromPath(path, mode);
+            return null;
+        }
+
+        private static void reset(LoadingMode mode) {
+            if (mode == LoadingMode.OVERRIDE) {
+                if (sdebug != null && props != null) {
+                    sdebug.println(
+                            "overriding other security properties files!");
+                }
+                props = new Properties() {
+                    @Override
+                    public synchronized Object put(Object key, Object val) {
+                        if (key instanceof String strKey &&
+                                val instanceof String strVal &&
+                                "include".equals(strKey)) {
+                            loadInclude(strVal);
+                            return null;
+                        }
+                        return super.put(key, val);
+                    }
+                };
+            }
+        }
+
+        private static void loadInclude(String propFile) {
+            String expPropFile = PropertyExpander.expandNonStrict(propFile);
+            if (sdebug != null) {
+                sdebug.println("processing include: '" + propFile + "'" +
+                        (propFile.equals(expPropFile) ? "" :
+                                " (expanded to '" + expPropFile + "')"));
+            }
+            try {
+                Path path = Path.of(expPropFile);
+                if (!path.isAbsolute()) {
+                    if (currentPath == null) {
+                        throw new InternalError("Cannot resolve '" +
+                                expPropFile + "' relative path when included " +
+                                "from a non-regular properties file " +
+                                "(e.g. HTTP served file)");
+                    }
+                    path = currentPath.resolveSibling(path);
+                }
+                loadFromPath(path, LoadingMode.APPEND);
+            } catch (IOException | InvalidPathException e) {
+                throw new InternalError("Unable to include '" + expPropFile +
+                        "'", e);
+            }
+        }
+
+        private static void loadFromPath(Path path, LoadingMode mode)
+                throws IOException {
+            boolean isRegularFile = Files.isRegularFile(path);
+            if (isRegularFile) {
+                path = path.toRealPath();
+            } else if (Files.isDirectory(path)) {
+                throw new IOException("Is a directory");
+            } else {
```

Comment:
When would this happen?

Reply:
This would happen if pointing to a directory from the _include_ directive or the `java.security.properties` System property (either with a local path or a file-based URL). For the `java.security.properties` case, pointing to a directory with a URL prefixed by `file://` was leading to unexpected and wrong results before.

Reply:
Oh, I meant the final `else`. What does it mean if a file is neither "regular" nor "directory"? Also I don't quite understand why one uses `toRealPath` and one uses `toAbsolutePath`. Is this related to resolving a symlink?

Reply:
Oh, I see now, _Linux_ pipes are examples of non-regular + non-directory at the same time, where `Path::toAbsolutePath()` works but `Path::toRealPath()` fails:

```
[fferrari@vmhost ~]$ jshell -<<<'System.out.println(Path.of("/dev/stdin").toAbsolutePath())'
/dev/stdin
[fferrari@vmhost ~]$ jshell -<<<'System.out.println(Path.of("/dev/stdin").toRealPath())'
Exception java.nio.file.NoSuchFileException: /dev/stdin
      at UnixException.translateToIOException (UnixException.java:92)
      at UnixException.rethrowAsIOException (UnixException.java:106)
      at UnixException.rethrowAsIOException (UnixException.java:111)
      at UnixPath.toRealPath (UnixPath.java:825)
      at (#1:1)
```

That change was introduced to avoid a regression, please see [this comment](https://github.com/openjdk/jdk/pull/16483#issuecomment-1881663782) (and the commit message).

Reply:
Thanks. So only real file has real path.